src/tools/rbd_wnbd/rbd_mapping.cc

```diff
@@ -0,0 +1,293 @@
+{
+  std::unique_lock l{shutdown_lock};
+
+  dout(5) << __func__ << ": removing RBD mapping: " << cfg.devpath << dendl;
+
+  int r = 0;
+  if (!cfg.persistent && saved_cfg_to_registry) {
+    dout(5) << __func__ << ": cleaning up non-persistent mapping: "
+            << cfg.devpath << dendl;
+    r = remove_config_from_registry(&cfg);
+    if (r) {
+      derr << __func__ << ": could not clean up non-persistent mapping: "
+           << cfg.devpath << dendl;
+    }
+  }
+
+  if (handler) {
+    handler->shutdown();
+    delete handler;
+    handler = nullptr;
+  }
+
+  if (watch_ctx) {
+    r = image.update_unwatch(watch_handle);
+    if (r < 0) {
+      derr << __func__ << ": update_unwatch failed with error: "
+           << cpp_strerror(r) << dendl;
+    }
+    delete watch_ctx;
+    watch_ctx = nullptr;
+  }
+
+  image.close();
+  io_ctx.close();
+
+  return r;
+}
+
+int RbdMapping::start()
+{
+  dout(10) << "initializing mapping" << dendl;
+  int r = init();
+  if (r < 0) {
+    return r;
+  }
+
+  dout(10) << "starting wnbd handler" << dendl;
+  r = handler->start();
+  if (r) {
+    r = r == ERROR_ALREADY_EXISTS ? -EEXIST : -EINVAL;
+    return r;
+  }
+
+  dout(10) << "setting up watcher" << dendl;
+  watch_ctx = new WNBDWatchCtx(io_ctx, handler, image, initial_image_size);
+  r = image.update_watch(watch_ctx, &watch_handle);
+  if (r < 0) {
+    derr << __func__ << ": update_watch failed with error: "
+         << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // Wait for the mapped disk to become available.
+  r = wait_mapped_disk(cfg);
+  if (r < 0) {
+    return r;
+  }
+
+  // We're storing mapping details in the registry even for non-persistent
+  // mappings. This allows us to easily retrieve mapping details such
+  // as the rbd pool or admin socket path.
+  // We're cleaning up the registry entry when the non-persistent mapping
+  // gets disconnected or when the ceph service restarts.
+  r = save_config_to_registry(&cfg);
+  if (r < 0) {
+    return r;
+  } else {
+    saved_cfg_to_registry = true;
+  }
+
+  if (disconnect_cbk) {
+    monitor_thread = std::thread([this]{
+      int ret = this->wait();
+      // Allow "this" to be destroyed by the disconnect callback.
+      this->monitor_thread.detach();
+      dout(5) << "finished waiting for: " << this->cfg.devpath
+              << ", ret: " << ret << dendl;
+      disconnect_cbk(this->cfg.devpath, ret);
+    });
+  }
+
+  return 0;
+}
+
+// Wait until the image gets disconnected.
+int RbdMapping::wait()
+{
+  if (handler) {
+    return handler->wait();
+  }
+  return 0;
+}
+
+RbdMapping::~RbdMapping()
+{
+  dout(10) << __func__ << ": cleaning up rbd mapping: "
+           << cfg.devpath << dendl;
+  shutdown();
+}
+
+// Wait for the mapped disk to become available.
+int wait_mapped_disk(Config& cfg)
+{
+  DWORD status = WnbdPollDiskNumber(
+    cfg.devpath.c_str(),
+    TRUE, // ExpectMapped
+    TRUE, // TryOpen
+    cfg.image_map_timeout * 1000,
```

Comment:
WnbdPollDiskNumber expects the timeout in milliseconds while cfg.image_map_timeout uses seconds:

https://github.com/cloudbase/wnbd/blob/20e029d04df111a1a4ef12e98130a73491dee6d4/include/wnbd.h#L189-L198
https://github.com/ceph/ceph/blob/93ace0d66e94d1f7b478776501acdc5267d05867/src/tools/rbd_wnbd/rbd_wnbd.h#L66

Reply:
So there was a bug here previously?

Reply:
Unfortunately yes. It went unnoticed since the default is 20 and it usually takes a matter of milliseconds for the disk to show up, also the downstream branch had this patch.

Reply:
This change didn't get included in Reef, so I'd say we can skip the tracker ticket: https://github.com/ceph/ceph/commit/a48b8d90d96667482d0fd8a747f1f0280c4fc2f2